accord-core/src/main/java/accord/coordinate/Barrier.java

```diff
@@ -0,0 +1,309 @@
+            }
+            else if (command.is(Status.Invalidated))
+                // Surface the invalidation and the barrier can be retried by the caller
+                Barrier.this.tryFailure(new Timeout(command.txnId(), command.homeKey()));
+
+        }
+
+        @Override
+        public PreLoadContext listenerPreLoadContext(TxnId caller)
+        {
+            return PreLoadContext.contextFor(caller);
+        }
+    }
+
+    private ExistingTransactionCheck checkForExistingTransaction()
+    {
+        checkState(seekables.size() == 1 && seekables.domain() == Domain.Key);
+        ExistingTransactionCheck check = new ExistingTransactionCheck();
+        Key k = seekables.get(0).asKey();
+        node.commandStores().mapReduceConsume(
+                contextFor(k),
+                k.toUnseekable(),
+                minEpoch,
+                Long.MAX_VALUE,
+                check);
+        return check;
+    }
+
+    // Hold result of looking for a transaction to act as a barrier for an Epoch
+    static class BarrierTxn
+    {
+        @Nonnull
+        public final TxnId txnId;
+        @Nonnull
+        public final Timestamp executeAt;
+        @Nonnull
+        public final Status status;
+        @Nonnull
+        public final Key key;
+        public BarrierTxn(@Nonnull TxnId txnId, @Nonnull Timestamp executeAt, @Nonnull Status status, Key key)
+        {
+            this.txnId = txnId;
+            this.executeAt = executeAt;
+            this.status = status;
+            this.key = key;
+        }
+
+        public BarrierTxn max(BarrierTxn other)
+        {
+            if (other == null)
+                return this;
+            return status.compareTo(other.status) >= 0 ? this : other;
+        }
+    }
+
+    /*
+     * Check for an existing transaction that is either already Applied (dependencies were applied)
+     * or PreApplied (outcome fixed, dependencies not yet applied).
+     *
+     * For Applied we can return success immediately with the executeAt epoch. For PreApplied we can add
+     * a listener for when it transitions to Applied and then return success.
+     */
+    class ExistingTransactionCheck extends AsyncResults.AbstractResult<BarrierTxn> implements MapReduceConsume<SafeCommandStore, BarrierTxn>
+    {
+        @Override
+        public BarrierTxn apply(SafeCommandStore safeStore)
+        {
+            BarrierTxn found = safeStore.mapReduceWithTerminate(
+                    seekables,
+                    safeStore.ranges().allSince(minEpoch),
+                    // TODO review this with Benedict, but I don't believe an exclusive sync point works here
+                    TestKind.Any,
+                    TestTimestamp.EXECUTES_AFTER,
+                    TxnId.minForEpoch(minEpoch),
+                    TestDep.ANY_DEPS,
+                    null,
+                    // TODO is PreApplied right? It's the first OutcomeKnown state, but why not Committed?
```

Comment:
is this TODO still relevant?

Reply:
I still have this question. I think it's correct I am just not 100% on why. Is `Committed` not good enough because `Committed` doesn't mean the outcome is known because it could be `Committed` at some nodes and `Invalidated` at others?

Reply:
No, if something is `Committed` it cannot be `Invalidated`. The state machine branches at `PreCommitted` - from here it must go to `Applied`. We otherwise skip straight from `AcceptedInvalidated` to `Invalidated` which is a terminal state.

Reply:
Ah so I could use `Committed` instead? I don't want to use anything that could still be `Invalidated` because then it might not wait on its dependencies.

I must have misunderstood what `OutcomeKnown` encompasses.

Reply:
Outcome refers to the `Writes` and `Result` registers, which are computed during the Execute phase.